dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/action/validation/BaseProgramRuleActionValidator.java

```diff
@@ -193,4 +189,93 @@ private ProgramRuleActionValidationResult handleTrackedEntityAttribute(
 
     return ProgramRuleActionValidationResult.builder().valid(true).build();
   }
+
+  public List<ProgramStage> getRequiredProgramStages(
+      ProgramRuleActionValidationContext validationContext, Program program) {
+    if (validationContext == null || program == null) {
+      return Collections.emptyList();
+    }
+
+    List<ProgramStage> availableStages = validationContext.getProgramStages();
+
+    if (availableStages == null) {
+      availableStages = new ArrayList<>();
+    }
+
+    if (hasAllProgramStages(availableStages, program)) {
+      return availableStages;
+    }
+
+    List<ProgramStage> missingStages =
+        fetchMissingStages(validationContext, program, availableStages);
+
+    return combineStages(availableStages, missingStages);
+  }
+
+  private boolean hasAllProgramStages(List<ProgramStage> availableStages, Program program) {
+    if (availableStages == null || program == null || program.getProgramStages() == null) {
+      return false;
+    }
+
+    Set<String> availableStageIds =
+        availableStages.stream()
+            .filter(Objects::nonNull)
+            .map(ProgramStage::getUid)
+            .collect(Collectors.toSet());
+
+    Set<String> allProgramStageIds =
+        program.getProgramStages().stream()
+            .filter(Objects::nonNull)
+            .map(ProgramStage::getUid)
+            .collect(Collectors.toSet());
+
+    return availableStageIds.containsAll(allProgramStageIds);
+  }
+
+  private List<ProgramStage> fetchMissingStages(
+      ProgramRuleActionValidationContext validationContext,
+      Program program,
+      List<ProgramStage> availableStages) {
+    if (program.getProgramStages() == null) {
+      return Collections.emptyList();
+    }
+
+    Set<String> availableStageIds =
+        availableStages.stream()
+            .filter(Objects::nonNull)
+            .map(ProgramStage::getUid)
+            .collect(Collectors.toSet());
+
+    List<String> missingStageIds =
+        program.getProgramStages().stream()
+            .filter(Objects::nonNull)
+            .map(ProgramStage::getUid)
+            .filter(stageId -> !availableStageIds.contains(stageId))
+            .toList();
+
+    if (missingStageIds.isEmpty()) {
+      return Collections.emptyList();
+    }
+
+    ProgramStageService stageService =
+        validationContext.getProgramRuleActionValidationService().getProgramStageService();
+
+    return stageService.getProgramStages(missingStageIds);
+  }
+
+  private List<ProgramStage> combineStages(
+      List<ProgramStage> availableStages, List<ProgramStage> missingStages) {
+    List<ProgramStage> combined = new ArrayList<>(availableStages);
+
+    if (missingStages != null && !missingStages.isEmpty()) {
+      Set<String> existingIds =
```

Comment:
This is a private method and you created `missingStages` getting the uids that were not present in `availableStages`, do you need to filter again the duplicates?
Simply concatenating the lists would be enough here